{
 "metadata": {
  "name": "",
  "signature": "sha256:a86e0524297bb9dfb9bde074d91ead86088dc49b01bd1bdaf5e348581b355fdf"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "# Genome Composition in Hymenoptera\n",
      "\n",
      "Interval loci (iLoci) provide a mechanism for breaking down an annotated genome sequence into manageable chunks, each representing a gene (or set of overlapping genes), an intergenic region, or an incompletely assembly fragment of the genome. This document uses iLoci to explore genome composition in 12 Hymenopteran species.\n",
      "\n",
      "## Data inputs\n",
      "\n",
      "Genome annotations (in GFF3 format) were collected from NCBI for 11 of the 12 species. Annotations for the 12th species (*Polistes dominula*) were produced in-house and are pending publication.\n",
      "\n",
      "The [LocusPocus program](http://aegean.readthedocs.org/en/v0.10.2/locuspocus.html) was used to process each annotation and produce a new GFF3 file for each species containing iLocus annotations. These GFF3 files are the inputs to this procedure."
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Unannotated sequences\n",
      "\n",
      "LocusPocus currently doesn't report iLoci for unannotated genomic sequences. To make sure we account for the entire genome sequence, we analyze all `##sequence-region` pragmas and note those with no associated annotations. This function determines the length of a sequence based on its `##sequence-region` pragma (see the [GFF3 spec](http://www.sequenceontology.org/gff3.shtml))."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import re\n",
      "\n",
      "def seqlen(pragmastring):\n",
      "    \"\"\"\n",
      "    Determines the length of a genomic sequence from its corresponding GFF3 \n",
      "    pragma record.\n",
      "    \"\"\"\n",
      "    pattern = \"##sequence-region\\s+(\\S+)\\s+(\\d+)\\s+(\\d+)\"\n",
      "    seqmatch = re.search(pattern, pragmastring)\n",
      "    assert seqmatch, \"unable to parse sequence region pragma: %s\" % pragmastring\n",
      "\n",
      "    seqid = seqmatch.group(1)\n",
      "    start = int(seqmatch.group(2))\n",
      "    end   = int(seqmatch.group(3))\n",
      "    assert end >= start, \"end < start: %s\" % pragmastring\n",
      "\n",
      "    seqlength = end - start + 1\n",
      "    return seqid, seqlength\n",
      "\n",
      "def test_seqlen():\n",
      "    \"\"\"\n",
      "    Unit tests for seqlen function. Throws an assertion error in case of\n",
      "    failure, so no output is good output!\n",
      "    \"\"\"\n",
      "    test_strings = { \"##sequence-region   chr1 1 1000000\":      1000000,\n",
      "                     \"##sequence-region\\tscaf0001   401   600\": 200,\n",
      "                     \"##sequence-region   2L 1001 10000\\n\":     9000 }\n",
      "    \n",
      "    for pragmastring in test_strings:\n",
      "        length = test_strings[pragmastring]\n",
      "        seqid, test_length = seqlen(pragmastring)\n",
      "        assert length == test_length, \"seqlen test failed: %s\" % pragmastring\n",
      "\n",
      "test_seqlen()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 12
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Overlapping iLoci\n",
      "\n",
      "In general iLoci are intended to be distinct non-overlapping genomic regions. However, in gene-dense regions there will occasionally be some overlap in the genes' regulatory regions. We do not want to count these overlapping regions twice when we're giving a breakdown of genome composition, so this function is used to collapse a set of overlapping iLoci into a single gene-containing region for record-keeping purposes."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from intervaltree import Interval\n",
      "\n",
      "def merge_loci(loci):\n",
      "    \"\"\"\n",
      "    Input is a set of interval objects, each corresponding to an iLocus.\n",
      "    Output is a new interval object containing all the iLoci.\n",
      "    \"\"\"\n",
      "    loci = list(loci)\n",
      "    start, end = 0, 0\n",
      "    for locus in loci:\n",
      "        start = min(start, locus.begin)\n",
      "        if start == 0:\n",
      "            start = locus.begin\n",
      "        end = max(end, locus.end)\n",
      "    return Interval(start, end, loci)\n",
      "\n",
      "def test_merge_loci():\n",
      "    \"\"\"\n",
      "    Unit tests for merge_loci function.\n",
      "    \"\"\"\n",
      "    testint = merge_loci((Interval(1, 3, \"dude\"),\n",
      "                          Interval(2, 4, \"sweet\")))\n",
      "    assert testint.begin == 1 and testint.end == 4\n",
      "    \n",
      "    testint = merge_loci((Interval(6, 9,  \"rad\"),\n",
      "                          Interval(7, 30, \"wicked\")))\n",
      "    assert testint.begin == 6 and testint.end == 30\n",
      "\n",
      "test_merge_loci()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 13
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## GFF3 parsing\n",
      "\n",
      "Here's where things get hairy.\n",
      "\n",
      "* **Step 1**: Walk through the GFF3 file line-by-line. If the line is a `##sequence-region` pragma, store the length of the corresponding sequence. If the line is an iLocus annotation, store it in an interval tree (for efficient interval queries) and discard the length computed from the `##sequence-region` pragma. Ignore all other lines in the GFF3 file.\n",
      "* **Step 2**: All sequence lengths remaining in memory correspond to unannotated sequences. Compute their sum and store.\n",
      "* **Step 3**: Iterate over all iLoci in the interval trees and add their lengths to the appropriate category (gene-containing, gene-less internal, or gene-less fragment). For iLoci that overlap other iLoci, verify the assumption that they all contain genes, and merge before computing length."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from intervaltree import IntervalTree\n",
      "from collections import namedtuple\n",
      "GenomeComposition = namedtuple(\"GenomeComposition\",\n",
      "                               \"genes internal terminal unannot\")\n",
      "def parse_gff3(fp):\n",
      "    \"\"\"\n",
      "    Input:  handle to GFF3 data (LocusPocus output)\n",
      "    Output: a simple `GenomeComposition` data structure giving a breakdown of\n",
      "            the number of nucleotides in each of 4 categories\n",
      "    \"\"\"\n",
      "    #----------\n",
      "    # Step 1\n",
      "    #----------\n",
      "    \n",
      "    # Dictionary of interval trees, indexed by sequence ID, for storing\n",
      "    # iLoci\n",
      "    loci = {}\n",
      "    \n",
      "    # Dictionary of sequence lengths, indexed by sequence ID\n",
      "    seqlengths = {}\n",
      "    \n",
      "    for line in fp:\n",
      "        if line.startswith(\"##sequence-region\"):\n",
      "            seqid, slength = seqlen(line)\n",
      "            seqlengths[seqid] = slength\n",
      "        elif \"\\tlocus\\t\" in line:\n",
      "            fields = line.rstrip().split(\"\\t\")\n",
      "            assert len(fields) == 9\n",
      "            seqid = fields[0]\n",
      "            start = int(fields[3])\n",
      "            end = int(fields[4]) + 1 # Python interval tree is [)\n",
      "            if seqid not in loci:\n",
      "                loci[seqid] = IntervalTree()\n",
      "            loci[seqid][start:end] = line\n",
      "            if seqid in seqlengths:\n",
      "                seqlengths.pop(seqid, None)\n",
      "\n",
      "\n",
      "    #----------\n",
      "    # Step 2\n",
      "    #----------\n",
      "    unannot = 0\n",
      "    for seqid in seqlengths:\n",
      "        unannot += seqlengths[seqid]\n",
      "\n",
      "\n",
      "    #----------\n",
      "    # Step 3\n",
      "    #----------\n",
      "    withgenes = 0\n",
      "    sansgenes = 0\n",
      "    ends = 0\n",
      "    merged_loci = {}\n",
      "    for seqid in loci:\n",
      "        merged_loci[seqid] = IntervalTree()\n",
      "        for locus in loci[seqid]:\n",
      "            if merged_loci[seqid][locus.begin:locus.end]:\n",
      "                continue\n",
      "\n",
      "            overlap = loci[seqid][locus.begin:locus.end]\n",
      "            assert len(overlap) > 0\n",
      "            if len(overlap) == 1:\n",
      "                locuslength = locus.end - locus.begin\n",
      "                if \";gene=\" in locus.data:\n",
      "                    withgenes += locuslength\n",
      "                else:\n",
      "                    sansgenes += locuslength\n",
      "                    if \";end=true\" in locus.data:\n",
      "                        ends += locuslength\n",
      "            else:\n",
      "                cont = True\n",
      "                while cont:\n",
      "                    count = len(overlap)\n",
      "                    for testlocus in overlap:\n",
      "                        assert \";gene=\" in testlocus.data\n",
      "                    merged_locus = merge_loci(overlap)\n",
      "                    overlap = loci[seqid][merged_locus.begin:merged_locus.end]\n",
      "                    cont = len(overlap) > count\n",
      "                merged_loci[seqid].add(merged_locus)\n",
      "                locuslength = merged_locus.end - merged_locus.begin\n",
      "                withgenes += locuslength\n",
      "    \n",
      "    return  GenomeComposition(withgenes, sansgenes - ends, ends, unannot)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 14
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Load data into a DataFrame\n",
      "\n",
      "Now that we've defined our GFF3 parsing procedure, let's go ahead and run it for each of the 12 species, and store the results in a data frame for analysis and visualization."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import pandas as pd\n",
      "\n",
      "# Dictionaries will end up as columns in the data frame\n",
      "withgenes = {}\n",
      "internal = {}\n",
      "terminal = {}\n",
      "unannot = {}\n",
      "withgenes_perc = {}\n",
      "sansgenes_perc = {}\n",
      "\n",
      "# Process the GFF3 files (takes a good 30-60 seconds)\n",
      "species_list = \"Ador Aflo Amel Bimp Bter Dmel Hsal Mrot Nvit Pdom Sinv Tcas\".split(\" \")\n",
      "for species in species_list:\n",
      "    infile = \"%s.iloci.gff3\" % species\n",
      "    with open(infile, \"r\") as fp:\n",
      "        comp = parse_gff3(fp)\n",
      "        withgenes[species] = comp.genes\n",
      "        internal[species] = comp.internal\n",
      "        terminal[species] = comp.terminal\n",
      "        unannot[species] = comp.unannot\n",
      "        total = comp.genes + comp.internal + comp.terminal + comp.unannot\n",
      "        withgenes_perc[species] = float(comp.genes) / float(total)\n",
      "        sansgenes_perc[species] = 1.0 - withgenes_perc[species]\n",
      "\n",
      "# Create and print the data frame\n",
      "# Hat tip: http://www.gregreda.com/2013/10/26/intro-to-pandas-data-structures/\n",
      "dfdata = { \"Genes\": withgenes, \"Internal\": internal, \"Terminal\": terminal,\n",
      "           \"Unannotated\": unannot, \"PercGenic\": withgenes_perc,\n",
      "           \"PercNongenic\": sansgenes_perc }\n",
      "compdata = pd.DataFrame(dfdata)\n",
      "compdata"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>Genes</th>\n",
        "      <th>Internal</th>\n",
        "      <th>PercGenic</th>\n",
        "      <th>PercNongenic</th>\n",
        "      <th>Terminal</th>\n",
        "      <th>Unannotated</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>Ador</th>\n",
        "      <td> 145955249</td>\n",
        "      <td>  52236431</td>\n",
        "      <td> 0.633651</td>\n",
        "      <td> 0.366349</td>\n",
        "      <td> 23789448</td>\n",
        "      <td>  8359043</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>Aflo</th>\n",
        "      <td> 111534743</td>\n",
        "      <td>  88987308</td>\n",
        "      <td> 0.483949</td>\n",
        "      <td> 0.516051</td>\n",
        "      <td> 16615620</td>\n",
        "      <td> 13330110</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>Amel</th>\n",
        "      <td> 164647446</td>\n",
        "      <td>  65892048</td>\n",
        "      <td> 0.657878</td>\n",
        "      <td> 0.342122</td>\n",
        "      <td>  6977844</td>\n",
        "      <td> 12753319</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>Bimp</th>\n",
        "      <td> 145196994</td>\n",
        "      <td>  69612165</td>\n",
        "      <td> 0.582687</td>\n",
        "      <td> 0.417313</td>\n",
        "      <td> 23292044</td>\n",
        "      <td> 11083853</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>Bter</th>\n",
        "      <td> 151054900</td>\n",
        "      <td>  83382829</td>\n",
        "      <td> 0.607490</td>\n",
        "      <td> 0.392510</td>\n",
        "      <td>  4849573</td>\n",
        "      <td>  9366942</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>Dmel</th>\n",
        "      <td>  90106688</td>\n",
        "      <td>  30037032</td>\n",
        "      <td> 0.748509</td>\n",
        "      <td> 0.251491</td>\n",
        "      <td>   237826</td>\n",
        "      <td>        0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>Hsal</th>\n",
        "      <td>  83083255</td>\n",
        "      <td> 171227361</td>\n",
        "      <td> 0.282149</td>\n",
        "      <td> 0.717851</td>\n",
        "      <td> 29975084</td>\n",
        "      <td> 10179901</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>Mrot</th>\n",
        "      <td> 112248303</td>\n",
        "      <td> 111361241</td>\n",
        "      <td> 0.411678</td>\n",
        "      <td> 0.588322</td>\n",
        "      <td> 27589094</td>\n",
        "      <td> 21461931</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>Nvit</th>\n",
        "      <td> 124542814</td>\n",
        "      <td>  92489803</td>\n",
        "      <td> 0.418423</td>\n",
        "      <td> 0.581577</td>\n",
        "      <td> 29004658</td>\n",
        "      <td> 51611102</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>Pdom</th>\n",
        "      <td>  64321922</td>\n",
        "      <td> 119973287</td>\n",
        "      <td> 0.309201</td>\n",
        "      <td> 0.690799</td>\n",
        "      <td> 15654101</td>\n",
        "      <td>  8076910</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>Sinv</th>\n",
        "      <td>  50840266</td>\n",
        "      <td> 221768097</td>\n",
        "      <td> 0.128382</td>\n",
        "      <td> 0.871618</td>\n",
        "      <td> 43559529</td>\n",
        "      <td> 79841277</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>Tcas</th>\n",
        "      <td>  80514791</td>\n",
        "      <td> 115606404</td>\n",
        "      <td> 0.382950</td>\n",
        "      <td> 0.617050</td>\n",
        "      <td>  6134233</td>\n",
        "      <td>  7993305</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 17,
       "text": [
        "          Genes   Internal  PercGenic  PercNongenic  Terminal  Unannotated\n",
        "Ador  145955249   52236431   0.633651      0.366349  23789448      8359043\n",
        "Aflo  111534743   88987308   0.483949      0.516051  16615620     13330110\n",
        "Amel  164647446   65892048   0.657878      0.342122   6977844     12753319\n",
        "Bimp  145196994   69612165   0.582687      0.417313  23292044     11083853\n",
        "Bter  151054900   83382829   0.607490      0.392510   4849573      9366942\n",
        "Dmel   90106688   30037032   0.748509      0.251491    237826            0\n",
        "Hsal   83083255  171227361   0.282149      0.717851  29975084     10179901\n",
        "Mrot  112248303  111361241   0.411678      0.588322  27589094     21461931\n",
        "Nvit  124542814   92489803   0.418423      0.581577  29004658     51611102\n",
        "Pdom   64321922  119973287   0.309201      0.690799  15654101      8076910\n",
        "Sinv   50840266  221768097   0.128382      0.871618  43559529     79841277\n",
        "Tcas   80514791  115606404   0.382950      0.617050   6134233      7993305"
       ]
      }
     ],
     "prompt_number": 17
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Genic vs non-genic\n",
      "\n",
      "Perhaps the simplest breakdown of genome composition is genic vs non-genic. The code below borrows heavily from [this page](http://nbviewer.ipython.org/github/cs109/content/blob/master/lec_03_statistical_graphs.ipynb)."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%matplotlib inline\n",
      "import brewer2mpl\n",
      "import numpy as np\n",
      "import matplotlib.pyplot as plt\n",
      "from matplotlib import rcParams\n",
      "\n",
      "def remove_border(axes=None, top=False, right=False, left=True, bottom=True):\n",
      "    \"\"\"\n",
      "    Minimize chartjunk by stripping out unnecesasry plot borders and axis ticks\n",
      "    \n",
      "    The top/right/left/bottom keywords toggle whether the corresponding plot border is drawn\n",
      "    \"\"\"\n",
      "    ax = axes or plt.gca()\n",
      "    ax.spines['top'].set_visible(top)\n",
      "    ax.spines['right'].set_visible(right)\n",
      "    ax.spines['left'].set_visible(left)\n",
      "    ax.spines['bottom'].set_visible(bottom)\n",
      "    \n",
      "    #turn off all ticks\n",
      "    ax.yaxis.set_ticks_position('none')\n",
      "    ax.xaxis.set_ticks_position('none')\n",
      "    \n",
      "    #now re-enable visibles\n",
      "    if top:\n",
      "        ax.xaxis.tick_top()\n",
      "    if bottom:\n",
      "        ax.xaxis.tick_bottom()\n",
      "    if left:\n",
      "        ax.yaxis.tick_left()\n",
      "    if right:\n",
      "        ax.yaxis.tick_right()\n",
      "\n",
      "ticks = range(len(species_list))\n",
      "ticks = [i+1.5 for i in ticks]\n",
      "plt.bar(range(1, len(species_list)+1), compdata[\"PercGenic\"], color=\"#B2182B\", label=\"Genic\")\n",
      "plt.bar(range(1, len(species_list)+1), compdata[\"PercNongenic\"], bottom=compdata[\"PercGenic\"], color=\"#2166AC\", label=\"Non-genic\")\n",
      "plt.xticks(ticks, species_list, rotation=\"horizontal\")\n",
      "plt.ylabel(\"Proportion of genome\")\n",
      "plt.xlabel(\"\")\n",
      "remove_border()\n",
      "plt.show()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "display_data",
       "png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAEACAYAAAC08h1NAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAGq5JREFUeJzt3Xm8XVV99/HP1zDKKELBYjAiEQEJUwlUpmPVNqWPQC2I\nEZlaHlCfABa1DviY+PJ5FFpECqiEKUytsUCUUAlYmZUhDCEECBaUMYgCokBkSMivf6x1kp3Duffu\ne3PXTe7d3/frdV93T2et395nn99ZZ+1JEYGZmTXHm1Z2AGZmNrSc+M3MGsaJ38ysYZz4zcwaxonf\nzKxhnPjNzBqmaOKXdL6k30ia18syp0t6SNJcSTuVjMfMzMq3+KcBE3qaKWlfYKuIGAscDXyvcDxm\nZo1XNPFHxM3A870ssh9wYV72dmBDSZuWjMnMrOlWdh//5sATlfEngbevpFjMzBphZSd+AHWM+x4S\nZmYFrbaS618AjK6Mvz1PW46kmDx58tLxVqtFq9UatCCkzu+eFdN5/yOXv/LKH+yyS5c/kra9y++7\n/CHQdQVUOhBJY4ArI2L7LvP2BSZFxL6SdgdOi4jduywXJeOUxDuPvmxQynrk7AO77jwuf+WUP5hl\nly5/pG17l993+UOga+Iv2uKX9H1gH2BjSU8Ak4HVASJiakRcJWlfSQ8DC4EjS8ZjZmaFE39ETKyx\nzKSSMZiZ2fJWhYO7ZmY2hJz4zcwaxonfzKxhnPjNzBrGid/MrGGc+M3MGsaJ38ysYZz4zcwaxonf\nzKxhnPjNzBrGid/MrGGc+M3MGsaJ38ysYZz4zcwaxonfzKxhnPjNzBrGid/MrGGc+M3MGsaJ38ys\nYZz4zcwaxonfzKxhnPjNzBrGid/MrGGc+M3MGsaJ38ysYZz4zcwaxonfzKxhnPjNzBrGid/MrGGc\n+M3MGsaJ38ysYZz4zcwaxonfzKxhnPjNzBrGid/MrGGc+M3MGqZo4pc0QdKDkh6S9IUu8zeWdLWk\neyTdJ+mIkvGYmVnBxC9pFHAmMAHYFpgoaZuOxSYBcyJiR6AFfEvSaqViMjOzsi3+8cDDEfFoRCwC\npgP7dyzza2D9PLw+8FxELC4Yk5lZ45VsXW8OPFEZfxLYrWOZc4DrJD0FrAd8tGA8ZmZG2cQfNZb5\nMnBPRLQkvQv4L0k7RMSLnQtOmTJl6XCr1aLVag1WnGZmjVIy8S8ARlfGR5Na/VXvA/4/QET8UtIj\nwNbAnZ2FVRO/mZkNXMk+/juBsZLGSFoDOBiY2bHMg8AHASRtSkr6vyoYk5lZ4xVr8UfEYkmTgGuA\nUcB5ETFf0jF5/lTgG8A0SXNJX0L/FBG/KxWTmZmV7eohImYBszqmTa0MPwt8uGQMZma2PF+5a2bW\nME78ZmYN48RvZtYwTvxmZg3jxG9m1jBO/GZmDePEb2bWML4Fso14j5x94MoOwWyV4sRvI97MzbYb\ntLL2e/r+QSvLbGWp1dUj6c2Sti4djJmZlddn4pe0HzCHdM8dJO0kqfNma2ZmNkzUafFPIT1A5XmA\niJgDbFkwJjMzK6hO4l8UEb/vmLakRDBmZlZenYO790s6BFhN0ljgOOCWsmGZmVkpdVr8xwLbAa8C\n3wdeAD5TMigzMyunzxZ/RCwkPRv3y+XDMTOz0vpM/JJ2JSX9MZXlIyLGFYzLzMwKqdPH/2/A54D7\n8EFdM7Nhr07ifyYifN6+mdkIUSfxf03SecBPgdfytIiIGeXCMjOzUuok/sOBrfOy1a4eJ34zs2Go\nTuL/M+A9ERGlgzEzs/LqnMd/C7Bt6UDMzGxo1Gnx/zlwj6RHSBdxgU/nNDMbtuok/gn5f7urR4Vi\nMTOzIdBnV09EPApsCOwHfBjYIE8zM7NhqM79+I8HLgE2ATYFLpF0XOnAzMysjDpdPUcBu+V79iDp\nJOA24PSSgVn/+LmyZlZX3WfuLulh2FYRg/VcWT9T1mzkq5P4pwG3S5pBOrB7AHB+0ajMzKyYOrdl\nPlXSjcCepDN7jsiPXzQzs2GoblfPPcDTefmQtEVEPF4uLDMzK6XO/fiPBSYDvwVer8zavlRQK4MP\njppZU9Rp8X8G2DoinisdzMrkg6Nm1hR17tXzOOk5u2ZmNgLUafE/Alwv6ccsfz/+U/t6oaQJwGnA\nKODciDi5yzIt4NvA6sCzEdGqF7qZmQ1EncT/eP5bI//VImkUcCbwQWABcIekmRExv7LMhsB3gL+K\niCclbdyf4M3MrP/qnM45BUDSOu2rd2saDzzcvq+PpOnA/sD8yjIfBy6PiCdzXc/2o3wzMxuAOmf1\nvA84F1gPGC1pB+CYiPh0Hy/dHHiiMv4ksFvHMmOB1SVdn8v/14i4uG7wNnL4rCqzoVOnq+c00q2Z\nrwCIiLmS9qnxujpP7Fod2Bn4APBm4FZJt0XEQzVeayOIz6oyGzq1LuCKiMel5W7Dv7jGyxYAoyvj\no0mt/qonSAd0XwZelnQTsAPwhsQ/ZcqUpcOtVotWq1UndDMz61Dr4K6kPQAkrQEcx/L99D25Exgr\naQzwFHAwMLFjmSuAM/OB4DVJXUFdzxaqJn4zMxu4Oon/U8C/kvrsFwA/Af5PXy+KiMWSJgHXkE7n\nPC8i5ks6Js+fGhEPSroauJd0189zIuKBga2KmZnVUeesnmdIZ9/0W0TMAmZ1TJvaMX4KcMpAyjcz\ns/6rc1bPGaQDtWLZAdsXgDsi4oqCsdkqxGfdmI0cdbp61gK2Bi4lJf+/I13NO07S+yPiMwXjs1WE\nz7oxGznqJP5xwB4RsRhA0neBn5Huzz+vYGxmZlZAnZu0bQisWxlfF9gofxG8UiQqMzMrpk6L/5+B\nOZJuIHX17AN8Q9I6wE8LxmZmZgXUOavnPEmzSPfeCeDEiFiQZ3++ZHBmZjb46l65+xTwo8KxmJnZ\nEKjTx29mZiNIj4lf0juHMhAzMxsavbX4LwOQdN0QxWJmZkOgtz7+UZJOBN4t6QTSGT1ttR69aGZm\nq57eWvwfA14n3WBtvfy3bmXYzMyGoR5b/BHxIHCSpHsj4qohjMnMzAqqc1bPLZK+Lemu/PctSRsU\nj8zMzIqok/jPJ92N8yDgo8CLwLSSQZmZWTl1LuB6V0R8pDI+RdLcUgGZmVlZdVr8L0vaqz0iaU/g\nj+VCMjOzkuq0+D8JXFTp138eOLxcSGZmVlKdm7TdQ3roygZ5/A/FozIzs2Jq3aQNnPDNzEYK36TN\nzKxhnPjNzBqmVlePpD2AMZXlIyIuKhXUSPTI2Qeu7BDMzIAaiV/SJcCWwD2ke/e0OfH3w8zNthuU\ncvZ7+v5BKcfMmqtOi38XYNuIiNLBmJlZeXX6+O8D3lY6EDMzGxp1WvybAA9Img28mqdFROxXLiwz\nMyulTuKfkv+3u3pUGTYzs2GmzpW7N0jaDNiVlPBnR8Rvi0dmZmZF9NnHL+mjwO0suy3zbEkHlQ7M\nzMzKqNPV8xVg13YrX9ImwLXApSUDMzOzMuqc1SPgmcr4cyz/4HUzMxtG6rT4rwaukfTvpIR/MDCr\naFRmZlZMncT/T8BHgD1JB3enRsQPi0ZlZmbF1DmrJ4DL85+ZmQ1zPfbxS/p5/v+SpBc7/l4YuhDN\nzGww9djij4g98v91B1q4pAnAacAo4NyIOLmH5XYFbgU+GhEzBlqfmfWf7xzbPHXuznlxRBza17Qu\nrxsFnAl8EFgA3CFpZkTM77LcyaSDyD5byGyI+c6xzVPn4O57qyOSViPdsbMv44GHI+LR/LrpwP7A\n/I7ljgUuI10ZbGYd3CK3wdZj4pf0ZeBLwNqSXqzMWgScXaPszYEnKuNPArt11LE56cvgL1h2Swgz\nq3CL3AZbjwd3I+IbwAbARRGxXuVvo4j4Yo2y6yTx04Av5jOHhLt6zMyK67WrJyKWSBo/wLIXAKMr\n46NJrf6qXYDpkgA2Bv5a0qKImNlZ2JQpU5YOt1otWq3WAMMyM2u2On38d0kaHxGz+1n2ncBYSWOA\np0hX/E6sLhARW7aHJU0DruyW9GH5xG9mZgNXJ/HvDnxC0mPAwjwtImJcby+KiMWSJgHXkE7nPC8i\n5ks6Js+fugJxm5nZANVJ/H+V/1cfxFJLRMyi474+PSX8iDiybrlmZkNlJJ5VVeeWDY9K2hHYi5T8\nb46IucUjMzNbBYzEs6rqPIjleOAS0rN3NwUukXRc6cDMzKyMOl09RwG7RcRCAEknAbcBp5cMzMzM\nyqjzIBaAJT0Mm5nZMFOnxT8NuF3SDNKB3QOA84tGZWZmxdQ5uHuqpBuBPfKkIyJiTtmwzMyslLpd\nPbDsNE7fVsHMbBirc1bPV4ELgI1It1WYJun/Fo7LzMwKqdPH/wlgXES8AiDpm8Bc4OslAzMzszLq\ndPUsANaujK/FG2+2ZmZmw0SdFv8LwP2SfpLHPwTMlnQG6Z49vpjLzGwYqZP4f5j/2vfquSEPCz84\nxcxs2KlzOucFktYE3p0nPRgRi8qGZWZmpdR52HoLuBB4LE/aQtLhEXFjycDMzKyMOl09pwJ/GRG/\nAJD0bmA6sHPJwMzMrIw6Z/Ws1k76ABHx39T7wjAzs1VQ3Ucvnku6NbOAQ0iPVTQz69NIfJDJcFcn\n8X8SmAS0T9u8GfhusYjMbEQZiQ8yGe56TfySVgPmRsR7gG8NTUhmZlZSr338EbEY+IWkdwxRPGZm\nVlidrp6NSFfuzgYW5mkREfuVC8vMzEqpk/i/kv9Xb8fsK3bNzIapHhO/pLVJB3a3Au4FzvcVu2Zv\n5LNWbLjprcV/IfAa6SyefYFtgeOHIiiz4cRnrdhw01vi3yYitgeQdB5wx9CEZGZmJfV2Vs/i9kA+\nu8fMzEaA3lr84yS9WBlfuzIeEbF+wbjMzKyQHhN/RIwaykDMzGxo1LlJm5mZjSBO/GZmDePEb2bW\nME78ZmYN48RvZtYwTvxmZg3jxG9m1jBO/GZmDVM88UuaIOlBSQ9J+kKX+YdImivpXkk/lzSudExm\nZk1WNPFLGgWcCUwg3d1zoqRtOhb7FbB3RIwDvg6cXTImM7OmK93iHw88HBGP5nv5Twf2ry4QEbdG\nxB/y6O3A2wvHZGbWaKUT/+bAE5XxJ/O0nvwDcFXRiMzMGq7OoxdXRO1HNEp6P/D3wB7d5k+ZMmXp\ncKvVotVqrWBoZmbNVDrxLwBGV8ZHk1r9y8kHdM8BJkTE890KqiZ+MzMbuNJdPXcCYyWNkbQGcDAw\ns7qApC2AGcAnIuLhwvGYmTVe0RZ/RCyWNAm4BhgFnBcR8yUdk+dPBb4KvAX4niSARRExvmRcZmZN\nVrqrh4iYBczqmDa1MnwUcFTpOMzMLPGVu2ZmDePEb2bWME78ZmYN48RvZtYwTvxmZg3jxG9m1jBO\n/GZmDePEb2bWME78ZmYN48RvZtYwTvxmZg1T/F49ZmYlPXL2gSs7hGHHid/MhrWZm203KOXs9/T9\ng1LOcOCuHjOzhnHiNzNrGCd+M7OGceI3M2sYJ34zs4Zx4jczaxgnfjOzhnHiNzNrGCd+M7OGceI3\nM2sYJ34zs4Zx4jczaxgnfjOzhnHiNzNrGCd+M7OGceI3M2sYJ34zs4Zx4jczaxgnfjOzhnHiNzNr\nGCd+M7OGceI3M2uYoolf0gRJD0p6SNIXeljm9Dx/rqSdSsZjZmYFE7+kUcCZwARgW2CipG06ltkX\n2CoixgJHA98rFc9gmvfqQpe/ksofzrG7fJe/qijZ4h8PPBwRj0bEImA6sH/HMvsBFwJExO3AhpI2\nLRjToJj3WuGdx+WvlLJdvstf1csfLCUT/+bAE5XxJ/O0vpZ5e8GYzMwar2Tij5rLaYCvMzOzAVBE\nmTwraXdgSkRMyONfApZExMmVZc4CboiI6Xn8QWCfiPhNR1n+MjAzG4CI6Gxcs1rB+u4ExkoaAzwF\nHAxM7FhmJjAJmJ6/KH7fmfShe+BmZjYwxRJ/RCyWNAm4BhgFnBcR8yUdk+dPjYirJO0r6WFgIXBk\nqXjMzCwp1tVjZmarJl+5WyHpAElLJG3dw/wbJO1Soi5Jm0i6XdJdkvaU9IikjQYr9gHEd4SkM/Lw\n65LmSLonx/fnefqfSrp0kOrrqY53SOrsIhxIufflsk+QNChdh5KmSPpsHn6pY97S7dfPMsdImtdl\n+hJJF1fGV5P0jKQr+1n+/p3X03TUcUpl/HOSJvdR3tck/UUe/oyktSvz2tt+nqT/qM6rLLN0G64o\nSSfm93lurne8pHN6Wt9+lv3WXOYcSb+W9GQevltSyS7zIpz4lzcR+E/eeCyiLejHWUeSetu+nXV9\nALg3InaJiJ/VraOX8lZUdT3/GBE7RcSOwJeAbwJExFMRcdAg1de1DuCdwMf7U1DHB7Fd7nuBDwF/\nDfSazPohehjuNr6iFgLbSVorj3+IdPrzG+rpIxH9LemCym5eA/5W0lvzeJ/rEBGTI+K6PHo88ObK\n7Pa23z6X/cluRfRVRx25ofA3wE4RsQPp8/RERPzviJi/ouVHxHN5XXYCzgJOzeM7R8TiFS1/qDnx\nZ5LWBXYjHWw+OE9bW9J0SQ9ImgFUWzMTJd2bWzMnVaa/JOkUSfcAu9epS9IOwMnA/rkFsVbH8ifk\neuZJOr5m7C1JN0r6kaRfSjpJ0qGSZue4t8zLbSLpsjx9tqT39bGpNgB+l1+7tHWaW7g/kvST/Gtl\nUm4x3i3pVklvycvdIOm0Sktw197qAE4C9srLHy/pTZL+Jcc6V9LRlfW9WdIVwP3dAo+IZ0hXiE/q\nZ8zvkjRL0p2SblK9X1VLf1VIOiiv6z2Sbqxsu5uUft0s/YXTh6tIyQ3SF/z32/XklvPFkn4GXKj0\nS+m6vI1+Kml0fm8/DPxL3p5bdpS/CDgb+MflVkRaX9KjlfF1JD2u9KvjAkl/J+lY4E+B6yVd2yX2\nnwFb5defKOkXkm4Gtq6Uu6Ok23LMMyRtmKffIOlUSXdImi9pV0k/lPTfkr6eX74Z8Gy+WJSI+F1E\n/Dq/dudczkuS/l9+H26V9CeSNuhh3Ub18V4ox3FLLu92Sev29L5Kelue3t7v9+yj/LIiwn/pOMch\nwFl5+CZgZ+AE4Nw8bXvSB2Nn0g7+GPBW0oHra4H983JLgAMHUNfhwBmVZR4BNgJ2Ae4lfemsA9wH\n7FijvBbwPLApsAawgHR6LcBxwLfz8L8De+ThLYAH8vAR7XiAxcAcYD7we2DnPH0MMK+y/EM5xo2B\nPwBH53mnAsfn4euBqXl4r8rrO+vYKU/fB7iysq5HAyfm4TWBO3IcLeAl4B0d2+bFLtv/eeBP+hHz\ntaRbi0D6gr02D08GPtsRf/vvMeD0PO9e4G15eP38f21gzTw8Frijc5t2rgdpH7w0r/ec6rYBpuRt\n0S7zSuDQPHwk8MM8PA34SA/75YvAeqR9b33gs8DkPO9HQCsPHwyc3Vleft1GnduedBLJFcAxLNuf\n18p1PQScUNlOe+Xhr7FsH70e+GZl332KZfv1E8Bb8ns4B/gF8B1g78pr2/vrEuBv8vDJLNuPuq5b\nL5/fycDngV8Cu+Rp65JyQU/v62eBL+dhAeuuzHznFv8yE0kfKvL/j5MS0yUAETGPtGMK2JV0/cFz\nEfE68G/A3vm1rwOX97OunrpnBOwJzIiIlyNiITAjx9VXeUHa6X4TEa8BD5POsIL05TEmD38QOFPS\nHNKHcz1J63SU/3Kkn7XbkO69dFEP8V4fEQsj4llS8m73P8+r1AeppUpE3AysL2n9LnW0+7M7++P/\nEjgsx3sb6ctxqzxvdkQ81kNsndpdDL3GnLfF+4BLc51nkVqXndrxt7sDvlqJ/eekVvhRLDuTbg3g\nXEn3Av9Bz90vywJO++AY0vv74y7rMzMiXs3ju5O+1CHtw9UWZo/HOCLiRdL7e1zHrB+Qf00CH8vj\nfVk7b7M7gEeB80n77oyIeCXXNRPSrwpgg7xPQLqVy96Vsmbm//cB91X2618BW+TPxi6khsEzwA8k\nHd4Rz2sR0d5ud7FsnxzIuq0J/Doi7gKIiJdyLuh8X9vHF2YDRyodMxkXES91K3SoDLuDEiUoHUR9\nP/BepYvFRpE+SHfT/UPS2S+pyrRXIn+t97Ourt0TeV41hmpdvZX3Y+DVyuuWVMaXsOy9F7Bb/hBV\n4+y6DhFxm6SNJW3cZXad+roW2486ACZFxH91xNsi9YP3KndvvB4Rzygd4+0r5jcBz+dk3h9L37OI\n+JSk8aRumruUThA4jpQ4Ds3dCq/ULHcmcAqptb9Jx7w/9hRDh7761U8j7fvTKtOuBL6Ru792Bq7r\n9sIOL3dut7xfde7P3XROr74vne/ZKICIWALcCNyo1AXZmfgXdbyuvU8OZN168o90eV8j4mZJewH/\nC7hA0qkRcXFvBZXkFn9yIHBRRIyJiHdGxBakn613kw8sSnovMI70oZkN7KN0pH8UqZVw4wrWNbrL\nsgHcDBygdLxhHeCAPK2v8vZ+Y3Fd/YRK607Sju3BbgtLeg/pg/ZczfI7yxLLjkPsSbpo78Uudbwp\n19Hufmi7Bvi08gFMSe+WVD2g2HMQ0iakFnuds20ES1vAj0g6MJchSePq1Fep910RMTsiJpNao6NJ\nXSlP50UOIyevGs4nddl1NhQ6369bSPslpK7Am/Lwi7nuHkXE86TW6j9Upr1EarmfTupe6vbl0WfZ\nOY4DJK0laT1SIiQiXgCer/R9Hwrc0EdZS+X9YGxl0k6k7rY+1Vy3Tq8Cb5P0Z7n+9XIu6Pq+StoC\neCYizgXOzfGtNG7xJx8jHUSsupz05qwt6QFS3/OdABHxtKQvkvoPBfxnRLS7CPraaXqq62TSDtEW\nua45ki4gfdkAnBMRc2uU9ylS90431bOTjgO+I2kuaX+4Efh0xzLtn+yQ1vewiIjcYo4uZdJluLrc\nK5LuzvX9fQ91HJ7rmAu8rnSwfBrpwzkGuFspgN+SzlTprL+tXe7qpH74iyLi1H7GfAjwPUlfyeV8\nn9Tt17n+VdXX/3NOSgJ+GhFzJX0XuFzSYcDVpOMT3eJYblpELCDd7ryzjs51ORaYJunzpG3Uvjhy\nOnCO0sHYgyLiVz3U+y3SQfDqtB+QvhBaXeKDdGD4akkLIuID3dYj788/AObmuGZXZh8OnJW/yH9J\n9ws6e3qf1wXOyAeEF5OOHRwDXNbD+nWW09e6dXqd1IA5Q+k01T+Suk17el/fD3xO0iLSF+RhNesp\nwhdw2ZCSdD3pgOjdKzsWs6ZyV4+ZWcO4xW9m1jBu8ZuZNYwTv5lZwzjxm5k1jBO/mVnDOPGbmTWM\nE7+ZWcP8D43J3YfSU6X3AAAAAElFTkSuQmCC\n",
       "text": [
        "<matplotlib.figure.Figure at 0x10ead4390>"
       ]
      }
     ],
     "prompt_number": 19
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}